stmt/stmts/select.go

```diff
@@ -149,7 +149,7 @@ func (s *SelectStmt) Plan(ctx context.Context) (plan.Plan, error) {
 		}
 	}
 	lock := s.Lock
-	if variable.ShouldAutocommit(ctx) {
+	if lock != coldef.SelectLockNone && ctx.(autocommit.Checker).ShouldAutocommit(ctx) {
```

Comment:
I think you can add BindAutoCommitChecker(ctx, checker) and let session register it explicitly like other context mechanism does , and then you can use autocommit.ShouldAutoCommit(ctx) directly.